database/migrations/2019_12_17_024553_add_foreign_keys.php

```diff
@@ -0,0 +1,623 @@
+use App\Models\Account\ImportJobReport;
+use App\Models\Account\ActivityStatistic;
+use App\Models\Relationship\Relationship;
+use Illuminate\Database\Schema\Blueprint;
+use App\Models\Relationship\RelationshipType;
+use Illuminate\Database\Migrations\Migration;
+use App\Models\Relationship\RelationshipTypeGroup;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+
+class AddForeignKeys extends Migration
+{
+    private $existingAccounts;
+    private $existingUsers;
+    private $existingContacts;
+
+    /**
+     * Add foreign keys to all the tables that don't have ones.
+     * Before adding foreign keys, in all the tables, we check whether the
+     * foreign keys actually have data from the table they point to. This will
+     * ensure data integrity from now on.
+     * In order to do this, we need to parse a lot of contacts, accounts and
+     * users, as most data point to them somehow. For those 3 models, when we
+     * launch the script, we put all those ids in arrays. Then, for each
+     * foreign key, we check in the arrays if the key exists, instead of querying
+     * the database again and again. This is much faster.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::disableForeignKeyConstraints();
+
+        $this->initialize();
+        $this->cleanActivityStatisticTable();
+        $this->cleanCallsTable();
+        $this->cleanContactTagTable();
+        $this->cleanDayTable();
+        $this->cleanDebtTable();
+        $this->cleanEntriesTable();
+        $this->cleanGenderTable();
+        $this->cleanGiftTable();
+        $this->cleanImportJobReportTable();
+        $this->cleanImportJobTable();
+        $this->cleanInvitationTable();
+        $this->cleanJournalEntryTable();
+        $this->cleanModuleTable();
+        $this->cleanNoteTable();
+        $this->cleanNotificationTable();
+        $this->cleanPetTable();
+        $this->cleanRelationshipTypeGroupTable();
+        $this->cleanRelationshipTypeTable();
+        $this->cleanRelationshipTable();
+        $this->cleanSpecialDateTable();
+        $this->cleanTagTable();
+        $this->cleanTaskTable();
+        $this->cleanTermUserTable();
+        $this->cleanUserTable();
+    }
+
+    private function initialize()
+    {
+        $rows = DB::table('contacts')->select('id')->get();
+        $this->existingContacts = [];
+        foreach ($rows as $row) {
+            $this->existingContacts[$row->id] = 1;
+        }
+
+        $rows = DB::table('users')->select('id')->get();
+        $this->existingUsers = [];
+        foreach ($rows as $row) {
+            $this->existingUsers[$row->id] = 1;
+        }
+
+        $rows = DB::table('accounts')->select('id')->get();
+        $this->existingAccounts = [];
+        foreach ($rows as $row) {
+            $this->existingAccounts[$row->id] = 1;
+        }
+    }
+
+    private function contactExistOrFail(int $id)
+    {
+        if (isset($this->existingContacts[$id])) {
```

Comment:
I use `isset` for performance reasons. This is the fastest way to iterate through an array.

Reply:
👍